pkg/services/sqlstore/user.go

```diff
@@ -172,18 +217,26 @@ func (ss *SQLStore) GetUserByLogin(ctx context.Context, query *models.GetUserByL
 
 		// Try and find the user by login first.
 		// It's not sufficient to assume that a LoginOrEmail with an "@" is an email.
-		user := &models.User{Login: query.LoginOrEmail}
-		has, err := sess.Where(notServiceAccountFilter(ss)).Get(user)
+		user := &models.User{}
+		where := "login=?"
+		if ss.Cfg.CaseInsensitiveLogin {
+			where = "LOWER(login)=LOWER(?)"
+		}
 
+		has, err := sess.Where(notServiceAccountFilter(ss)).Where(where, query.LoginOrEmail).Get(user)
 		if err != nil {
 			return err
 		}
 
 		if !has && strings.Contains(query.LoginOrEmail, "@") {
```

Comment:
nit: unrelated to this PR, but the logic in this method seems overcomplicated to me - why not just check `where login=? or email=?`? The only difference seems to be that we're checking that email contains "@", but do we really care about it? Or do we want to accommodate for cases when one user's login is another user's email 🤔

Reply:
user's login is another user's email is an issue that already exists in grafana. 
 
By checking login first, we assure we always prioritize login over email match (this could be optimized with a single query and a asc)

The user would always need to have valid basic auth, except in the case of advanced auth (but the attacker would need to control the idp or AD). But it could be worth investigating and sanitizing in a separate issue